### PR TITLE
[ICD] Add ICD CheckInSender class

### PR DESCRIPTION
**Pull request by jepenven-silabs**
Break-out PR for the ICD Check-In message feature.

This class was tested along with the complete changes manually with an All-cluster-app -icd build, Chip-tool and Wireshark to monitored the sending of the Check-in Message on the network.

Standard behavior is : 

- Create Object with all required parameters
- Call RequestResolve() with the FabricID and NodeID for the registered ICD Client
- Upon MDNS resolution the ICD Check-in message will be created and send on the network
- Upon MDNS failure, nothing will happen for said ICD Client until the next Active Time. No retries are needed since the MDNS resolution already have a couple retries embedded in it.

Since DNS resolution, and exchangeCtx message sending are already covered by other Unit tests, no Unit test were added for this class. Instead integration tests will be added once the feature is completed (ICD Check-In sending AND Receiving)

### CODE REVIEW

File: src/app/icd/ICDCheckInSender.cpp
Comment (mkardous-silabs) [inline src/app/icd/ICDCheckInSender.cpp:32]: Do we need a parameterized constructor when this is done by the Init function and the default constructor?

File: src/app/icd/ICDCheckInSender.cpp
Comment (mkardous-silabs) [inline src/app/icd/ICDCheckInSender.cpp:90]:
```suggestion
    CHIP_ERROR err = CHIP_NO_ERROR;
```
For good practice, we should not have an undefined variable.

File: src/app/icd/ICDCheckInSender.cpp
Comment (mkardous-silabs) [inline src/app/icd/ICDCheckInSender.cpp:102]: From all this, it seems we only need the `ExchangeManager`. It cannot be injected?

File: src/app/icd/ICDCheckInSender.cpp
Comment (mkardous-silabs) [inline src/app/icd/ICDCheckInSender.cpp:93]: We can't retreive it from the ICDManager and only have the todo in the ICDManager?

File: src/messaging/BUILD.gn
Comment (mkardous-silabs) [inline src/messaging/BUILD.gn:83]: I think this might brake the MRP increase based on polling intervals looking at the unit tests.

File: src/app/icd/ICDCheckInSender.h
Comment (mkardous-silabs) [inline src/app/icd/ICDCheckInSender.h:54]: I think we need a way to check that these were correctly set since we have multiple ways of injecting these objects.
We have no garantee that when we trigger a resolve for example that these values were set.

File: src/app/icd/BUILD.gn
Comment (mkardous-silabs) [inline src/app/icd/BUILD.gn:38]: can you brake this out of the source set and have another one for the check-in sender?

Comment (jepenven-silabs) [reply src/app/icd/BUILD.gn:38]: Nope, ICDManager will have to know about the ICD sender in a follow up PR

File: src/app/icd/ICDCheckInSender.cpp
Comment (jepenven-silabs) [reply src/app/icd/ICDCheckInSender.cpp:93]: Sure, for now let leave it as it is and inject it in the follow up PR.

File: src/app/icd/BUILD.gn
Comment (mkardous-silabs) [reply src/app/icd/BUILD.gn:38]: So add it as a dependency like what was done for all the other source sets?

File: src/app/icd/ICDCheckInSender.cpp
Comment (jepenven-silabs) [reply src/app/icd/ICDCheckInSender.cpp:32]: If we are using the `ObjectPool` (which we will in a future PR) then yes, this is needed.

File: src/messaging/BUILD.gn
Comment (jepenven-silabs) [reply src/messaging/BUILD.gn:83]: fixed with defines

File: src/app/icd/BUILD.gn
Comment (jepenven-silabs) [reply src/app/icd/BUILD.gn:38]: But why should the ICD  have more than 1 sources set ? When in fact these are required for the feature ?

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.cpp:69]: Compressed IDs are not guaranteed to be unique.  Ideally we would not be using compressed IDs here, and would instead track which fabric index we are working with to start with.

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.cpp:117]: This (and the sanity-checks above) would be a _lot_ more readable with the exchange manager stored in a temporary.

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.cpp:127]: This allocation can fail.  Why is it not being checked?

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.cpp:88]: Why are we doing this stack buffer that we then copy into a packetbuffer?  Why not just allocate a packetbuffer of the right size, then put the data directly into it?  Would be less space, less data copying, etc.

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.cpp:121]: This is wrong.  CreateUnauthenticatedSession() takes the MRP config of the _remote_ as an argument...  But it doesn't matter too much, since we are not doing MRP anyway.  Still, either use the MRP config from the resolve, or document that the value here does not matter, but then you might as well just `GetDefaultMRPConfig()`.  Because the "local" thing is definitely 100% wrong here.

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.cpp:139]: mCurrentNodeId is not read anywhere.  Why does it exist?

But we _should_ be saving the FabricIndex, per comments above.

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.cpp:141]: What guarantees that the ICD stays in active mode to listen for those lookup results?  I bet nothing does.  Followup needed to make this actually work and make sense; I doubt you want to try to tackle that in this PR.

File: src/app/icd/ICDCheckInSender.h
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.h:32]: This needs documentation on what the storage is expected to provide, exactly.

File: src/messaging/ReliableMessageProtocolConfig.cpp
Comment (bzbarsky-apple) [inline src/messaging/ReliableMessageProtocolConfig.cpp:73]: This file is not including the config headers that ensure these are defined.  It needs to, if it's going to use the config macros directly.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/icd/ICDCheckInSender.cpp
Comment (jepenven-silabs) [reply src/app/icd/ICDCheckInSender.cpp:69]: Considering where this function is called, we don't have access to the FabricIndex anymore and there is no easy way of retrieving it either. 

The only two solutions I see for your comments are either : 

- Modifying the `peerID` class to store the FabricIndex
- Keeping the Current MonitoringEntry inside the ICDCheckInSender (that will bypass the need of re-finding the monitoring entry inside the MonitoringTable.

I'll implement the second one since it seems a fair tradeof between RAM usage and code efficiency.

File: src/app/icd/ICDCheckInSender.cpp
Comment (jepenven-silabs) [reply src/app/icd/ICDCheckInSender.cpp:121]: thanks for the clarification. Will do.

File: src/app/icd/ICDCheckInSender.cpp
Comment (jepenven-silabs) [reply src/app/icd/ICDCheckInSender.cpp:141]: That's an excellent point, and you're right, I'll most probably tackle this once the whole sending of the Check-In is completed (a.k.a after the next PR.)

File: src/app/icd/ICDCheckInSender.h
Comment (jepenven-silabs) [reply src/app/icd/ICDCheckInSender.h:32]: actually these can be optimized out

File: src/messaging/ReliableMessageProtocolConfig.cpp
Comment (jepenven-silabs) [reply src/messaging/ReliableMessageProtocolConfig.cpp:73]: done

File: src/app/icd/ICDCheckInSender.cpp
Comment (jepenven-silabs) [reply src/app/icd/ICDCheckInSender.cpp:127]: good point

File: src/app/icd/ICDCheckInSender.cpp
Comment (mkardous-silabs) [inline src/app/icd/ICDCheckInSender.cpp:78]:
```suggestion
    VerifyOrReturnError(mExchangeManager->GetSessionManager() != nullptr, CHIP_ERROR_INTERNAL)
```

**Review by mkardous-silabs — APPROVED** [review]

File: src/app/icd/ICDCheckInSender.cpp
Comment (mkardous-silabs) [reply src/app/icd/ICDCheckInSender.cpp:87]: Solution suggestion for the follow; leveraging the active mode reqs to force the ICD to stay in active mode.

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.cpp:43]: Might be more readable as:
```suggestion
    ReturnOnFailure(SendCheckInMsg(mCurrentEntry, result.address));
    ChipLogError(AppServer, "Failed to send the ICD Check-In message");
```

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.cpp:55]:
```suggestion
```

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.cpp:70]: This crashes if CreateUnauthenticatedSession returns a NullOptional.  Is that intended?

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.cpp:64]: So we are storing the whole entry so we can get its key, right?

Would it make sense to just store the fabric index  and look up the entry here?   Or just store the key?  An ICDMonitoringEntry is pretty large (100+ bytes), no?

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.cpp:87]: Please file an issue about needing to stay active here, and have a TODO comment referencing it.

**Review by bzbarsky-apple — APPROVED** [review]
Approving so this is not blocked on me, but storing the whole entry in the sender seems wrong.

File: src/app/icd/ICDCheckInSender.cpp
Comment (jepenven-silabs) [reply src/app/icd/ICDCheckInSender.cpp:64]: true, for now just storing the key will be much lighter.

File: src/app/icd/ICDCheckInSender.cpp
Comment (jepenven-silabs) [reply src/app/icd/ICDCheckInSender.cpp:43]: if we do it that way, we won't have the failure log if the SendCheckInMsg fails no? since we would already have return before the ChipLogError...

File: src/app/icd/ICDCheckInSender.cpp
Comment (jepenven-silabs) [reply src/app/icd/ICDCheckInSender.cpp:70]: fixed

File: src/app/icd/ICDCheckInSender.cpp
Comment (jepenven-silabs) [reply src/app/icd/ICDCheckInSender.cpp:87]: issue created

File: src/platform/Linux/BUILD.gn
Comment (andy31415) [inline src/platform/Linux/BUILD.gn:113]: why? this was not needed before. Something went wrong here.

Comment (andy31415) [reply src/platform/Linux/BUILD.gn:113]: may be related to #30527

File: src/messaging/BUILD.gn
Comment (andy31415) [reply src/messaging/BUILD.gn:83]: This remove was not ok: ICDManager is included by the CPP files, so you cannot just remove the dependency.

See #30527

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [reply src/app/icd/ICDCheckInSender.cpp]: Ah, yes, you're right.

File: src/app/icd/ICDCheckInSender.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDCheckInSender.cpp:71]: This comment needed to move to where it's actually relevant...

Comment (jepenven-silabs) [reply src/app/icd/ICDCheckInSender.cpp:71]: #30659

**Pull request merged**
